### PR TITLE
[Mono AOT] Fix error when returning zero sized struct

### LINKED ISSUE #103628: [ios-8.0/9.0] Cannot compile project with AOT enabled (Assertion at mini-llvm.c, condition `cinfo->ret.nslots == 1' not met)
### Description

I'm developing a large project and at the very beginning the code compiled without problems, but after a while I checked again and got this error, I don’t have time to look for which specific type breaks the assembly, is it possible to display a more specific error message and how can I debug it?

I tried the latest net9.0-preview 5 but the result is the same.

### Reproduction Steps

I can't provide the source code as this is a commercial product.

### Expected behavior

No compilation errors

### Actual behavior

Mono Ahead of Time compiler - compiling assembly ..../obj/Release/net9.0-ios/ios-arm64/linked/MugenMvvm.dll
AOTID 0D0A964D-FAF6-A1E4-FADC-1F08E0767345
* Assertion at /Users/runner/work/1/s/src/mono/mono/mini/mini-llvm.c:1605, condition `cinfo->ret.nslots == 1' not met

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### LINKED ISSUE #106259: [MonoAOT] [x64] Return non-void in Function of void return type
Found while fixing https://github.com/dotnet/runtime/issues/103628

```
Found return instr that returns non-void in Function of void return type!
  ret void undef, !dbg !23
 void
 ```

Mono IR:
```
*** Unoptimized LLVM IR for HelloWorld.Program:GetS1 () ***
define hidden monocc void @HelloWorld_Program_GetS1() #2 !dbg !4 {
BB0:
  %vreg_loc_19 = alloca %S1, align 1, !dbg !8
  %vreg_loc_16 = alloca %S1, align 1, !dbg !8
  %ret_buf = alloca ptr, align 8, !dbg !8
  br label %INIT_BB1, !dbg !8

INIT_BB1:                                         ; preds = %BB0
  br label %INITED_BB2, !dbg !8

INITED_BB2:                                       ; preds = %INIT_BB1
  br label %BB2, !dbg !8

BB2:                                              ; preds = %INITED_BB2
  call void @llvm.memset.p0.i32(ptr %vreg_loc_19, i8 0, i32 0, i1 false), !dbg !9
  br label %BB1, !dbg !9

BB1:                                              ; preds = %BB2
  call void @llvm.memset.p0.i32(ptr %ret_buf, i8 0, i32 ptrtoint (ptr getelementptr (void, ptr null, i32 1) to i32), i1 false), !dbg !8
  call void @llvm.memcpy.p0.p0.i32(ptr %ret_buf, ptr %vreg_loc_19, i32 0, i1 false), !dbg !8
  ret void undef, !dbg !8
}

***
```

### Steps to reproduce:
Build the following program with mono aot:
```
using System;
using System.Runtime.InteropServices;

[StructLayout(LayoutKind.Sequential)]
public struct S1
{
}

namespace HelloWorld
{
    internal class Program
    {
        public static S1 GetS1() => new S1();
        private static void Main(string[] args)
        {
            S1 s1 = GetS1();
        }
    }
}
```

Error is most likely caused by wrong logic here https://github.com/dotnet/runtime/blob/15e96faf4558b017ea8df1dc28d9b2169f0badc0/src/mono/mono/mini/mini-llvm.c#L1596

### PR DESCRIPTION
**Pull request by jkurdek**
This PR allows Mono AOT to handle returns of zero sized structs in registers. 

Example: 
```c#
public struct Foo
{
}
```
The `[StructLayout(LayoutKind.Sequential, Size = 1)]` attribute will be added during compilation [SharpLab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwHQCUCuAdgC4CWAtgKbYCSJlATgPYAOAyowG6kDGlAzgG5kyKAGZ0/Yg3w9i6AGJMmyAN7IAvsiA).

```c#
[StructLayout(LayoutKind.Sequential)] // or [StructLayout(LayoutKind.Auto)] 
public struct Foo
{
}
```

However, adding an explicit `[StructLayout(LayoutKind.Sequential]` or `[StructLayout(LayoutKind.Auto)]` will not generate any explicit size metadata causing the struct to have size 0 [SharpLab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwHQCUCuAdgC4CWAtgKbYCSJlATgPYAOAyowG6kDGlAzgG5kyANptiDfD2IAZAIYBPJvmIAKBctUBpUoWDYOAR3yUSpeQBsAlAF1kUAMzp+k6cXQAxJk2QBvZABfZCA==).

---

#### Mono Codegen:

```c#
[StructLayout(LayoutKind.Sequential)]
public struct Foo
{
}

public class Bar
{
   public static Foo GetFoo() => new Foo();
}
```

When processing the return of the `GetFoo()` method mono will evaluate size of struct `Foo` to be equal to zero. Then it tries to pass it in registers, calculating that it needs `0` registers to do so. 
https://github.com/dotnet/runtime/blob/bedf340098a3cf494f200d99a9a53d3df1d1c6c7/src/mono/mono/mini/mini-arm64.c#L1708

Mono JIT handles this by omitting codegen part when `cinfo->ret.nregs` is equal to zero.
https://github.com/dotnet/runtime/blob/d606c601510c1a1a28cb6ef3550f12db049c0776/src/mono/mono/mini/mini-arm64.c#L3894-L3911

This PR proposes to fix the issue in Mono AOT by treating zero sized structs as if they have size 1. 

---

Fixes https://github.com/dotnet/runtime/issues/103628
Fixes https://github.com/dotnet/runtime/issues/106259

### CODE REVIEW

**Comment by lambdageek** [comment]
> This PR proposes to fix the issue in Mono AOT by treating zero sized structs as if they have size 1.

Won't this break scenarios where we might call an AOTed method from the JIT or vice-versa?  Or does this not leak out into callers?

What happens if you set the ret type to void? do we get ill-formed LLVM IR because we end up trying to return a value?

**Comment by jkurdek** [comment]
> Won't this break scenarios where >we might call an AOTed method >from the JIT or vice-versa? Or >does this not leak out into >callers?

I'm not sure I get it. The change is made only on llvm codegen level. The size of the struct is not overridden for any other calls. Without this change we would just encounter compile time error. Am I missing something?

I don't think void return type would even enter this code path. Will verify.

File: src/mono/mono/mini/mini-llvm.c [hunk 1]
```diff
@@ -1610,7 +1610,11 @@ sig_to_llvm_sig_full (EmitContext *ctx, MonoMethodSignature *sig, LLVMCallInfo *
 	case LLVMArgVtypeAsScalar: {
 		int size = mono_class_value_size (mono_class_from_mono_type_internal (rtype), NULL);
 		/* LLVM models this by returning an int */
-		if (size < TARGET_SIZEOF_VOID_P) {
+		if (size == 0) {
+			/* Empty struct with LayoutKind attribute and without specified size */
+			g_assert(cinfo->ret.nslots == 0);
+			ret_type = LLVMIntType (8);
```
Comment (lambdageek) [inline src/mono/mono/mini/mini-llvm.c:1616]: > I don't think void return type would even enter this code path. Will verify.

I meant could you change it like this:

```suggestion
			ret_type = LLVMVoidType ();
```

**Review by lambdageek — APPROVED** [review]

File: src/mono/mono/mini/mini-llvm.c (hunk 1 above)
Comment (jkurdek) [reply src/mono/mono/mini/mini-llvm.c:1616]: Yeah. I tried this initially and it resulted in ill formed llvm ir.

**Comment by jkurdek** [comment]
Failures seem to be unrelated.

**Comment by derjabkin** [comment]
Is this PR also going to be merged into version 8.0?

**Comment by vitek-karas** [comment]
@jkurdek  when you have some time, please open a backport, I'll try to get it through the triage. I think it should be doable.

**Pull request merged**